sklearn/metrics/tests/test_dist_metrics.py

```diff
@@ -242,3 +247,20 @@ def custom_metric(x, y):
     pyfunc = DistanceMetric.get_metric("pyfunc", func=custom_metric)
     eucl = DistanceMetric.get_metric("euclidean")
     assert_array_almost_equal(pyfunc.pairwise(X), eucl.pairwise(X) ** 2)
+
+
+def test_readonly_kwargs():
+    # Non-regression test for:
+    # https://github.com/scikit-learn/scikit-learn/issues/21685
+
+    rng = check_random_state(0)
+
+    weights = rng.rand(100)
+    VI = rng.rand(10, 10)
+    weights.setflags(write=False)
+    VI.setflags(write=False)
```

Comment:
Are you sure this test fails on `main`?

<del>Setting numpy-level readonly flags is not as strong as allocating buffer memory in readonly mode at the OS level. I think your should use the `create_memmap_backed_data` with `mmap_mode="r"` here to make sure this is a proper non-regression test for #21685.</del>

Edit: I tried to run this test against `main` and it's indeed a valid non-regression test, I reproduce the following failure:

```python-traceback
________________________________________________ test_readonly_kwargs _________________________________________________

    def test_readonly_kwargs():
        # Non-regression test for:
        # https://github.com/scikit-learn/scikit-learn/issues/21685
    
        rng = check_random_state(0)
    
        weights = rng.rand(100)
        VI = rng.rand(10, 10)
        weights.setflags(write=False)
        VI.setflags(write=False)
    
        # Those distances metrics have to support readonly buffers.
>       DistanceMetric.get_metric("seuclidean", V=weights)

sklearn/metrics/tests/test_dist_metrics.py:259: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
sklearn/metrics/_dist_metrics.pyx:264: in sklearn.metrics._dist_metrics.DistanceMetric.get_metric
    return metric(**kwargs)
sklearn/metrics/_dist_metrics.pyx:447: in sklearn.metrics._dist_metrics.SEuclideanDistance.__init__
    self.vec = np.asarray(V, dtype=DTYPE)
stringsource:658: in View.MemoryView.memoryview_cwrapper
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   ValueError: buffer source array is read-only

stringsource:349: ValueError
```

which seems close enough to the original problem reported in #21685.